modules/standard/Socket.chpl

```diff
@@ -0,0 +1,1403 @@
+  }
+  else {
+    var len:socklen_t = buflen;
+    var buffer = c_calloc(c_uchar, buflen);
+    var err_out = sys_getsockopt(socketFd, level, optname, buffer:c_void_ptr, len);
+    if err_out != 0 {
+      c_free(buffer);
+      throw SystemError.fromSyserr(err_out, "Failed to get socket option");
+    }
+    return createBytesWithOwnedBuffer(buffer, len, buflen);
+  }
+}
+
+/*
+  Returns the value of the given socket option which is expected to be of type
+  :mod:`bytes <Bytes>` on provided :type:`tcpConn`. The needed symbolic constants (SO_* etc.)
+  are defined in :mod:`Sys` module.
+
+  :arg socket: socket to set option on
+  :type socket: `tcpConn`
+  :arg level: protocol level
+  :type level: `int(32)`
+  :arg optname: option to set.
+  :type optname: `int(32)`
+  :return: value of socket option
+  :rtype: :mod:`bytes <Bytes>`
+  :throws SystemError: Upon incompatible arguments
+                        and socket.
+*/
+proc getSockOpt(ref socket: tcpConn, level: c_int, optname: c_int, buflen: uint(16)): bytes throws {
+  var socketFd = socket.socketFd;
+  return getSockOpt(socketFd, level, optname, buflen) ;
+}
+
+/*
+  See Above
+*/
+proc getSockOpt(ref socket: tcpListener, level: c_int, optname: c_int, buflen: uint(16)): bytes throws {
+  var socketFd = socket.socketFd;
+  return getSockOpt(socketFd, level, optname, buflen) ;
+}
+
+/*
+  See Above
+*/
+proc getSockOpt(ref socket: udpSocket, level: c_int, optname: c_int, buflen): bytes throws {
+  var socketFd = socket.socketFd;
+  return getSockOpt(socketFd, level, optname, buflen) ;
+}
+
+pragma "no doc"
+proc getpeername(socketFD: fd_t) throws {
+  var addressStorage = new sys_sockaddr_t();
+  var err = sys_getpeername(socketFD, addressStorage);
+  if err != 0 {
+    throw SystemError.fromSyserr(err, "Failed to get remote address");
+  }
+  return new ipAddr(addressStorage);
+}
+
+/*
+  Returns the remote addressto which socket is connected. This is
+  useful to find out the port number of a remote IPv4/v6 socket, for instance.
+
+  :arg socket: socket to set option on
+  :type socket: `tcpConn`
+
+  :return: remote address
+  :rtype: `ipAddr`
+  :throws SystemError: If socket is not connected
+*/
+proc getpeername(ref socket: tcpConn): ipAddr throws {
+  var socketFd = socket.socketFd;
+  return getpeername(socketFd);
+}
+
+pragma "no doc"
+proc getsockname(socketFD: fd_t) throws {
+  var addressStorage = new sys_sockaddr_t();
+  var err = sys_getsockname(socketFD, addressStorage);
+  if err != 0 {
+    throw SystemError.fromSyserr(err, "Failed to get local address");
+  }
+  return new ipAddr(addressStorage);
+}
+
+/*
+  Returns the socket's own address. This is useful to find out the port
+  number of a remote IPv4/v6 socket, for instance.
+
+  :arg socket: socket to set option on
+  :type socket: `tcpConn`
+
+  :return: remote address
+  :rtype: `ipAddr`
+  :throws SystemError: If socket is closed.
+*/
+proc getsockname(ref socket: tcpConn): ipAddr throws {
```

Comment:
Same. Especially considering that we have `getSockOpt`, `getsockname` feels really wrong to me.